sensor/common/compliance/node_inventory_handler_impl.go

```diff
@@ -73,41 +80,84 @@ func (c *nodeInventoryHandlerImpl) ProcessMessage(_ *central.MsgToSensor) error
-func (c *nodeInventoryHandlerImpl) run() <-chan *central.MsgFromSensor {
-	toC := make(chan *central.MsgFromSensor)
-	go func() {
-		defer c.stopper.Flow().ReportStopped()
-		defer close(toC)
-		for {
-			select {
-			case <-c.stopper.Flow().StopRequested():
+func (c *nodeInventoryHandlerImpl) run() (<-chan *central.MsgFromSensor, <-chan *common.MessageToComplianceWithAddress) {
+	toCentral := make(chan *central.MsgFromSensor)
+	toCompliance := make(chan *common.MessageToComplianceWithAddress)
+
+	go c.nodeInventoryHandlingLoop(toCentral, toCompliance)
+
+	return toCentral, toCompliance
+}
+
+func (c *nodeInventoryHandlerImpl) nodeInventoryHandlingLoop(toCentral chan *central.MsgFromSensor, toCompliance chan *common.MessageToComplianceWithAddress) {
+	defer c.stopper.Flow().ReportStopped()
+	defer close(toCentral)
+	defer close(toCompliance)
+	for {
+		select {
+		case <-c.stopper.Flow().StopRequested():
+			return
+		case inventory, ok := <-c.inventories:
+			if !ok {
+				c.stopper.Flow().StopWithError(errInputChanClosed)
 				return
-			case inventory, ok := <-c.inventories:
-				if !ok {
-					c.stopper.Flow().StopWithError(errInputChanClosed)
-					return
-				}
-				if !c.centralReady.IsDone() {
-					// TODO(ROX-13164): Reply with NACK to compliance
-					log.Warnf("Received node inventory but Central is unavailable")
-					continue
-				}
-				if inventory == nil {
-					log.Warnf("Received nil node inventory: not sending to Central")
-					break
-				}
-				if nodeID, err := c.nodeMatcher.GetNodeID(inventory.GetNodeName()); err != nil {
-					log.Warnf("Node '%s' unknown to sensor: not sending node inventory to Central", inventory.GetNodeName())
-				} else {
-					inventory.NodeId = nodeID
-					metrics.ObserveReceivedNodeInventory(inventory)
-					log.Debugf("Mapping node inventory name '%s' to Node ID '%s'", inventory.GetNodeName(), nodeID)
-					c.sendNodeInventory(toC, inventory)
-				}
+			}
+			if !c.centralReady.IsDone() {
+				log.Warnf("Received NodeInventory but Central is not reachable. Requesting Compliance to resend NodeInventory later")
+				c.sendNackToCompliance(toCompliance, inventory)
+				continue
+			}
+			if inventory == nil {
+				log.Warnf("Received nil node inventory: not sending to Central")
+				break
+			}
+			if nodeID, err := c.nodeMatcher.GetNodeID(inventory.GetNodeName()); err != nil {
+				log.Infof("Sending NACK to compliance after receiving unknown NodeInventory with ID %s", inventory.GetNodeId())
+				c.sendNackToCompliance(toCompliance, inventory)
+			} else {
+				inventory.NodeId = nodeID
+				metrics.ObserveReceivedNodeInventory(inventory)
+				log.Debugf("Mapping NodeInventory name '%s' to Node ID '%s'", inventory.GetNodeName(), nodeID)
+				c.sendNodeInventory(toCentral, inventory)
+				log.Debugf("Sending ACK to compliance for NodeInventory with ID %s", inventory.GetNodeId())
+				c.sendAckToCompliance(toCompliance, inventory)
```

Comment:
This must be removed before merging - we used that only for testing the communication

Reply:
Done and added a TODO for the followup regarding removing the Nack to have another look here